orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -706,13 +827,39 @@ Status ToCheckpointState(gsl::span<const uint8_t> checkpoint_bytes, CheckpointSt
                     "Checkpoint is invalid. Checkpoint version ", checkpoint_version, " is not supported.");
 
   const auto* fbs_module_state = fbs_checkpoint->module_state();
+
+  fbs::utils::ExternalDataReader external_data_reader = nullptr;
+  std::optional<std::ifstream> external_data_stream;
+
+  if (fbs_module_state->has_external_data()) {
+    ORT_RETURN_IF_NOT(checkpoint_path.has_value(),
+                      "External data is present in the checkpoint but the checkpoint path is not provided. External data with loading from buffer is not supported yet.");
+    auto data_path = ExternalCheckpointDataPath(checkpoint_path.value());
+    external_data_stream = std::ifstream(data_path, std::ios::binary);
+
+    char errbuf[256];
+    ORT_RETURN_IF(external_data_stream.value().fail(),
+                  "Failed to open checkpoint's external data file: ", ToUTF8String(data_path), strerror_s(errbuf, sizeof(errbuf), errno));
```

Comment:
Don't forget you need to add separators/labels for the different values so the output is readable.

```suggestion
                  "Failed to open checkpoint's external data file. file:", ToUTF8String(data_path), " error:", strerror_s(errbuf, sizeof(errbuf), errno));
```